tiledb/sm/cpp_api/ndrectangle.h

```diff
@@ -0,0 +1,229 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the C++ API for the TileDB NDRectangle.
+ */
+
+#ifndef TILEDB_CPP_API_NDRECTANGLE_H
+#define TILEDB_CPP_API_NDRECTANGLE_H
+
+#include "context.h"
+#include "deleter.h"
+#include "dimension.h"
+#include "domain.h"
+#include "tiledb.h"
+#include "type.h"
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace tiledb {
+
+class NDRectangle {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /**
+   * Constructor
+   *
+   * @param ctx The TileDB context
+   * @param domain The input Domain
+   */
+  explicit NDRectangle(const tiledb::Context& ctx, const tiledb::Domain& domain)
+      : ctx_(ctx) {
+    tiledb_ndrectangle_t* capi_ndrect;
+    ctx.handle_error(tiledb_ndrectangle_alloc(
+        ctx.ptr().get(), domain.ptr().get(), &capi_ndrect));
+    ndrect_ = std::shared_ptr<tiledb_ndrectangle_t>(capi_ndrect, deleter_);
+  }
+
+  /**
+   * Constructor
+   *
+   * @param ctx The TileDB context
+   * @param ndrect The NDRectangle C API pointer
+   *
+   * @note The object will manage the lifetime of tiledb_ndrectangle_t*
+   */
+  NDRectangle(const tiledb::Context& ctx, tiledb_ndrectangle_t* ndrect)
+      : ctx_(ctx) {
+    ndrect_ = std::shared_ptr<tiledb_ndrectangle_t>(ndrect, deleter_);
+  }
+
+  NDRectangle(const NDRectangle&) = default;
+  NDRectangle(NDRectangle&&) = default;
+  NDRectangle& operator=(const NDRectangle&) = default;
+  NDRectangle& operator=(NDRectangle&&) = default;
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+
+  /**
+   * Adds an 1D range along a dimension name, in the form
+   * (start, end). The datatype of the range
+   * must be the same as the dimension datatype.
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * // Set an 1D range on dimension 0, assuming the domain type is int64.
+   * int64_t start = 10;
+   * int64_t end = 20;
+   * ndrect.set_range(0, start, end);
+   * @endcode
+   *
+   * @tparam T The dimension datatype.
+   * @param dim_name The name of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this NDRectangle.
+   */
+  template <class T>
```

Comment:
We probably need to specialize this and all similar APIs for strings.

Reply:
Yes, good catch I missed it.